candle-transformers/src/models/stable_diffusion/uni_pc.rs

```diff
@@ -0,0 +1,1048 @@
+                        .into_iter()
+                        .map(|f| f as usize)
+                        .collect(),
+                )
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum CorrectorConfiguration {
+    Disabled,
+    Enabled { skip_steps: HashSet<usize> },
+}
+
+impl Default for CorrectorConfiguration {
+    fn default() -> Self {
+        Self::Enabled {
+            skip_steps: [0, 1, 2].into_iter().collect(),
+        }
+    }
+}
+
+impl CorrectorConfiguration {
+    pub fn new(disabled_steps: impl IntoIterator<Item = usize>) -> Self {
+        Self::Enabled {
+            skip_steps: disabled_steps.into_iter().collect(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct UniPCSchedulerConfig {
+    /// Configure the UNIC corrector. By default it is disabled
+    pub corrector: CorrectorConfiguration,
+    /// Determines how sigma relates to a given timestep
+    pub sigma_schedule: SigmaSchedule,
+    /// Determines the points
+    pub timestep_schedule: TimestepSchedule,
+    /// The solver order which can be `1` or higher. It is recommended to use `solver_order=2` for guided
+    /// sampling, and `solver_order=3` for unconditional sampling.
+    pub solver_order: usize,
+    /// Prediction type of the scheduler function
+    pub prediction_type: PredictionType,
+    pub num_training_timesteps: usize,
+    /// Whether to use the "dynamic thresholding" method. This is unsuitable for latent-space diffusion models such
+    /// as Stable Diffusion.
+    pub thresholding: bool,
+    /// The ratio for the dynamic thresholding method. Valid only when `thresholding=True`.
+    pub dynamic_thresholding_ratio: f64,
+    /// The threshold value for dynamic thresholding.
+    pub sample_max_value: f64,
+    pub solver_type: SolverType,
+    /// Whether to use lower-order solvers in the final steps.
+    pub lower_order_final: bool,
+}
+
+impl Default for UniPCSchedulerConfig {
+    fn default() -> Self {
+        Self {
+            corrector: Default::default(),
+            timestep_schedule: TimestepSchedule::FromSigmas,
+            sigma_schedule: SigmaSchedule::Karras(Default::default()),
+            prediction_type: PredictionType::Epsilon,
+            num_training_timesteps: 1000,
+            solver_order: 2,
+            thresholding: false,
+            dynamic_thresholding_ratio: 0.995,
+            sample_max_value: 1.0,
+            solver_type: SolverType::Bh1,
+            lower_order_final: true,
+        }
+    }
+}
+
+impl SchedulerConfig for UniPCSchedulerConfig {
+    fn build(&self, inference_steps: usize) -> Result<Box<dyn Scheduler>> {
+        Ok(Box::new(EdmDpmMultistepScheduler::new(
+            self.clone(),
+            inference_steps,
+        )?))
+    }
+}
+
+struct Inner {
+    model_outputs: Vec<Option<Tensor>>,
+    lower_order_nums: usize,
+    order: usize,
+    last_sample: Option<Tensor>,
+}
+
+struct State(RefCell<Inner>);
```

Comment:
Could we avoid the `RefCell` here and expose the proper mutability requirements in the various methods instead?

Reply:
Yes, though it requires also changing the mutability requirements of `Scheduler::step`. I wanted to avoid changing any existing signatures since I'm not that familiar with the project, but this should be all that's needed: 73e9bc9c9e25ba789ee72a51b734179f4e4bc05f

Fwiw adding other multi-step schedulers like ipndm or dpm++ would most likely require this change as well.